### ISSUE TITLE
Problems with the gradient computation

### ISSUE THREAD

**Issue by manuelgpda** [qulacs/qulacs#659]
Dear devs,

First of all, thank you for your amazing work.

Recently I found an "issue" that is provided in the attached code snippets. 

Basically, the time taken for gradient computation can drastically change (a factor of 10^6) and even given different results when using the same circuit and the same function depending on how the gate is added. You can in principle add a parametric X gate (option 1), what is slow. Or you can add a multiqubit parametric Pauli rotation specified for a single qubit X. If the latter is done by defining a list `target` with all the qubits in the circuit and placing identities (this is `0`) in the list `pauli_id` (option 2), then the result is correct as expected and it is way faster than the single parametric gate. However, if the lists `target` and `pauli_id` are defined only with qubits on which the rotation is not acting with an identity (option 3), then  the time taken is almost as slow as for a single parametric gate but also if the rotation is made on a qubit that is not the first one, then the result can be wrong (check snippet 2). The rule seems to be that if one includes in both lists all the qubits `j` whose index is smaller than the one on which the rotation is applied, then the multi qubit parametric rotation is faster and gives the right result. I tried to see where the issue is coming from and I believe it is inside the CausalConeSimulator computation.

I believe this might also be related to previously reported issues like [this one](https://discuss.pennylane.ai/t/qualcs-device-and-qml-grad-timings/1717).

Snippet 1
```python
from qulacs import ParametricQuantumCircuit, GradCalculator, Observable
from time import time

n = 2
observable = Observable(n)
observable.add_operator(1.0, "X 0")
observable.add_operator(1.0, "X 1")
observable.add_operator(1.0, "Z 0 Z 1")

theta = 2.2

# option 1
t1 = time()
circuit = ParametricQuantumCircuit(n)
circuit.add_parametric_RX_gate(0, theta)
gcalc = GradCalculator()
result1 = gcalc.calculate_grad(circuit, observable)
t2 = time()
print("Option 1:",result1)
print("Time 1:",t2-t1)

# option 2
t3 = time()
target = [0]
pauli_ids = [1]
circuit = ParametricQuantumCircuit(n)
circuit.add_parametric_multi_Pauli_rotation_gate(target, pauli_ids, theta)
gcalc = GradCalculator()
result2 = gcalc.calculate_grad(circuit, observable)
t4 = time()
print("Option 2:",result2)
print("Time 2:",t4-t3)

# option 3
t5 = time()
target = [0,1]
pauli_ids = [1,0]
circuit = ParametricQuantumCircuit(n)
circuit.add_parametric_multi_Pauli_rotation_gate(target, pauli_ids, theta)
gcalc = GradCalculator()
result3 = gcalc.calculate_grad(circuit, observable)
t6 = time()
print("Option 3:",result3)
print("Time 3:",t6-t5)
```
```
Option 1: [(-0.8084964038195901+0j)]
Time 1: 14.150959968566895
Option 2: [(-0.8084964038195901+0j)]
Time 2: 11.57673692703247
Option 3: [(-0.8084964038195901+0j)]
Time 3: 0.00026416778564453125
```

Snippet 2
```python
n = 2
observable = Observable(n)
observable.add_operator(0.3, "X 0")
observable.add_operator(0.8, "X 1")
observable.add_operator(1.1, "Z 0 Z 1")

theta = 2.2

# option 1
t1 = time()
circuit = ParametricQuantumCircuit(n)
circuit.add_parametric_RX_gate(1, theta)
gcalc = GradCalculator()
result1 = gcalc.calculate_grad(circuit, observable)
t2 = time()
print("Option 1:",result1)
print("Time 1:",t2-t1)

# option 2
t3 = time()
target = [1]
pauli_ids = [1]
circuit = ParametricQuantumCircuit(n)
circuit.add_parametric_multi_Pauli_rotation_gate(target, pauli_ids, theta)
gcalc = GradCalculator()
result2 = gcalc.calculate_grad(circuit, observable)
t4 = time()
print("Option 2:",result2)
print("Time 2:",t4-t3)

# option 3
t5 = time()
target = [0,1]
pauli_ids = [0,1]
circuit = ParametricQuantumCircuit(n)
circuit.add_parametric_multi_Pauli_rotation_gate(target, pauli_ids, theta)
gcalc = GradCalculator()
result3 = gcalc.calculate_grad(circuit, observable)
t6 = time()
print("Option 3:",result3)
print("Time 3:",t6-t5)
```
```
Option 1: [(-0.8893460442015492+0j)]
Time 1: 14.029311895370483
Option 2: [(0.4079469997759596+0j)]
Time 2: 13.821652889251709
Option 3: [(-0.8893460442015492+0j)]
Time 3: 0.0003330707550048828

```

**Comment by KowerKoint** [comment]
Thank you for reporting.
We found `CaussalConeSimulator` and `QuantumGateBase::set_target_index_list`, which is called from `CausalConeSimulator` is wrong.
Please wait for a while to fix it.

**Comment by forest1040** [comment]
you can use backprop method.

```
# backprop
t7 = time()
circuit = ParametricQuantumCircuit(n)
circuit.add_parametric_RX_gate(1, theta)
# gcalc = GradCalculator()
# result1 = gcalc.calculate_grad(circuit, observable)
result4 = circuit.backprop(observable)
t8 = time()
print("backprop:",result4)
print("Time backprop:",t8-t7)
```

This is result.
```
backprop: [-0.8893460442015493]
Time backprop: 0.0001609325408935547
```

**Comment by KowerKoint** [comment]
The first problem: `GradCalculator` is sometimes too slow is fixed by #664 .
The second problem: `GradCalculator` sometimes calculates the wrong value is fixed by #661 .
